### PR TITLE
Tooltip: Fix positioning by anchoring to child element

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Alternative to #41176 

This PR updates the ToolTip component to ensure that its position is correctly relative to the first child of the ToolTip.

## Why?

Following on from #40740 where the Popover component was refactored to use the floatingUI library, it looks like the following issues were introduced:

* Tooltips were no longer centred relative to their child component
* Tooltips at the edge of the screen would overflow / get a horizontal scrollbar

It looks like it was also mentioned in this comment: https://github.com/WordPress/gutenberg/pull/40740#pullrequestreview-966004363

## How?

* Add a ref to the ToolTip component that gets passed to both the ToolTip's child, and as an anchorRef, used as a reference for the Popover — this ensures that the Popover has the correct reference point for calculating the Tooltip position.

* Set the min-width of the Tooltip to `min-content` instead of `0` to prevent horizontal scrollbars from appearing

## Testing Instructions

1. Test in Storybook and ensure that by default the `ToolTip` component is now positioned relative to its first child.
2. In the post editor, hover over the options (vertical ellipsis) at the top right, and check that the Tooltip does not overflow / introduce a scrollbar.
3. Hover over the Undo / Redo buttons and ensure the Tooltip is now positioned more closely to the expected position.
4. Insert a block, and hover over the controls in the block toolbar: the tooltips should now be centered to the button.

## Screenshots or screencast 

| Before | After |
| --- | --- |
| image | image |
|  | image |
| image | image |
| image | image |

### CODE REVIEW

File: packages/components/src/tooltip/index.js
Comment (andrewserong) [inline packages/components/src/tooltip/index.js]: This transformation is necessary because otherwise we end up with the following issue (tested in Storybook):

| Before | After |
| --- | --- |
| image | image |

I'm not _too_ sure if I've understood the switched `left` / `right` logic in the Popover, but so far in testing this function appears to be working correctly for the ToolTip to me. Very happy for feedback, though, if anyone has ideas for how to improve this 🙂

**Comment by andrewserong** [comment]
I've pinged a few folks who reviewed #40740. Apologies for the liberal use of pinging / adding reviewers here — I wasn't _too_ sure if the way I've gone about fixing this is the right approach, so just wanted to make sure y'all could see the PR just in case I missed some details about how best to work with the refactored Popover. Thanks!

File: packages/components/src/popover/index.js
Comment (youknowriad) [inline packages/components/src/popover/index.js]: Why is this necessary, I fear it would have an impact on the block toolbar...

File: packages/components/src/tooltip/index.js
Comment (youknowriad) [inline packages/components/src/tooltip/index.js]: Why do you think we need this here? Is the "positionToPlacement" in the "Popover" component not working properly? Should we fix that one instead?

File: packages/components/src/tooltip/index.js
Comment (ciampo) [reply packages/components/src/tooltip/index.js]: This [confused me](https://github.com/WordPress/gutenberg/pull/40740#discussion_r869488923) as well.

It's a pity that Popover and Tooltip would have "opposite" behaviour in terms of how `left` and `right` work.

File: packages/components/src/tooltip/index.js
Comment (ciampo) [inline packages/components/src/tooltip/index.js]: Should we add some more unit tests to make sure that this function works properly?

File: packages/components/src/tooltip/index.js
Comment (ciampo) [inline packages/components/src/tooltip/index.js]: Maybe we could add some inline code to explain why we're storing both `childRef` and `mergedChildRefs` (which are being used later in code in the `getElementWithPopover` function)

File: packages/components/src/tooltip/index.js
Comment (ciampo) [inline packages/components/src/tooltip/index.js]: Should we consider updating the APIs of `Tooltip` to match the ones of `Popover`? (i.e. adding the `placement` prop and deprecating the `position` prop)

**Review by ciampo — COMMENTED** [review]
Left a few inline comments, focusing more on the component in isolation and less on its integration in the editor (which I'm sure other folks can also help with).

I also realised that we also have [an experimental `Tooltip` component](https://github.com/WordPress/gutenberg/tree/e6e0ea8906c0fb01ea132e3386d5e42172614ba9/packages/components/src/ui/tooltip) — it would be interesting to assess if we should look into using that "new" implementation or discard it.

File: packages/components/src/tooltip/index.js
Comment (andrewserong) [reply packages/components/src/tooltip/index.js]: Yes, good idea, I'll add some tests tomorrow.

File: packages/components/src/tooltip/index.js
Comment (andrewserong) [reply packages/components/src/tooltip/index.js]: Added an inline comment to hopefully clear this one up a bit.

File: packages/components/src/tooltip/index.js
Comment (andrewserong) [reply packages/components/src/tooltip/index.js]: That's a good question, while digging into this, it seems that we need a little more smarts for the Tooltip positioning. "bottom left" means something different to "bottom-start", and in the case of the Tooltip — I think I'd be keen to retain `position` as it implies implicit offset logic, but potentially also expose `placement`. I'll write a separate longer comment with a couple of thoughts.

File: packages/components/src/popover/index.js
Comment (andrewserong) [reply packages/components/src/popover/index.js]: Good point, I thought maybe we could use some clearance between the edge of the screen and the Tooltip, but from looking at other web apps (e.g. Gmail) it looks like it's pretty common for Tooltips to go flush up against the viewport. I've removed this line.

Comment (andrewserong) [reply packages/components/src/popover/index.js]: Update: turns out without setting `padding` the Tooltip jumps around a bit when it hits the edge of the viewport:



I'll look into a fix tomorrow.

**Comment by andrewserong** [comment]
Thanks for the feedback @ciampo and @youknowriad! I've gotten a little closer to coming up with a slightly more thorough fix, with honing in some of the underlying issues. I've pushed a change in https://github.com/WordPress/gutenberg/pull/41268/commits/800194214aca11ea466fb6d83abbead4763fe0b4, but it needs some tidying up and for me to add some tests which I'll look into tomorrow.

> Is the "positionToPlacement" in the "Popover" component not working properly? Should we fix that one instead?

I spent some time digging into this question, and I think we have some subtle issues to tease apart. The main issue appears to be the gap between the idea of the placement point and the expectations of the existing position prop. In floatingUI, we set a placement point, and then the element extends in the opposite direction from that alignment. So, in the original Popover PR, flipping `left` and `right` makes sense when the Popover is large enough to extend into the area we think of when we use left/right. The logic isn't intuitive per se, but achieves the desired result, for the most part.

However, Tooltips, being quite small, have a slightly different expectation associated with them. I think for a Tooltip, when we say "bottom left" we probably mean something closer to "center the Tooltip at the bottom left corner of this element".

The good news is that we can pass a function to `offset` in floatingUI, so in my [latest commit](https://github.com/WordPress/gutenberg/commit/800194214aca11ea466fb6d83abbead4763fe0b4), I've used this functionality so that we can better approximate the expectation of position "bottom left" — in this case, [by setting `alignmentAxis`](https://floating-ui.com/docs/offset#alignmentaxis) to negative half the width of the Tooltip, we can place the Tooltip at the desired location. As a result, it means we can achieve placement like the following:

image

This has a couple of flow on implications if it seems like an okay approach:

* The Tooltip's `position` property is still useful as it implicitly switches on some centering offset logic, which means that I don't think we'd want to deprecate the property.
* The floatingUI functionality for `placement` is still super useful, and gives more control over the desired position of the Popover/Tooltip, so maybe we'd want to expose this as well as the `position` property?
* However, we'd then have two very similar properties.
* I believe we still want the transformation from `position` to `placement` in the `Tooltip` to be unique to the tooltip, and not the Popover, as the expectations surrounding how a Tooltip will be placed are different to a larger Popover containing arbitrary content. For example, when I tried to apply the logic in this PR (which tests pretty well for me so far), the same logic (use offset with alignment axis of half the floating width) looked very wrong for components like the publishing visibility or publish dates in the sidebar inspector.

Anyway, those are my thoughts from going down a bit of a floatingUI rabbit hole today. I'll tidy up the approach tomorrow if there aren't any objections, and then get the PR ready for review again. But, do let me know if you think there's a different direction I should be pursuing 🙂

**Comment by andrewserong** [comment]
> I also realised that we also have [an experimental Tooltip component](https://github.com/WordPress/gutenberg/tree/e6e0ea8906c0fb01ea132e3386d5e42172614ba9/packages/components/src/ui/tooltip) — it would be interesting to assess if we should look into using that "new" implementation or discard it.

It looks like that one uses Reakit for the Tooltip, rather than Popover. It'd probably be good to consolidate on the one approach, so I imagine if/when this PR lands, we might be able to look at a refactor? (But let me know if you think we should do it the other way around)

**Comment by ciampo** [comment]
> Anyway, those are my thoughts from going down a bit of a floatingUI rabbit hole today. I'll tidy up the approach tomorrow if there aren't any objections, and then get the PR ready for review again. But, do let me know if you think there's a different direction I should be pursuing 🙂

Thank you for sharing these thoughts. I think that we shouldn't have both `position` and `placement` as it would be quite confusing.  i'd love it if we could have consistency in the APIs of `Popover` and Tooltip`, but I also understand that the expectations for the 2 components can be different.

> It looks like that one uses Reakit for the Tooltip, rather than Popover. It'd probably be good to consolidate on the one approach, so I imagine if/when this PR lands, we might be able to look at a refactor? (But let me know if you think we should do it the other way around)

Makes sense — it'd be great to clean this duplication up

**Comment by andrewserong** [comment]
Part of the issue in this PR (minimum width of the popover) will be resolved by #41361, if that PR looks viable. I'll put this PR on hold in lieu of that PR, then can rebase this one to see about a possibly more minimal change just for the Tooltip.

**Comment by andrewserong** [comment]
The main issue to resolve with this PR has now been merged separately in https://github.com/WordPress/gutenberg/pull/41361 (the Tooltip being cut off at the edge of the viewport). The other idea in this PR (to tweak the position of the Tooltip so that it's anchored correctly to the child element) would still be good to get in, but is lower priority, so I'm going to set this PR aside for the moment while I work on some higher priority tasks, and then will 🤞 get back to this next week.

Thanks again for the 👀, all!

**Comment by afercia** [comment]
Thanks for looking into this. Just wanted to note that this is also an accessibility issue.

Right now, on trunk, the tooltip may appear on top of some input fields, making impossible for keyboard users to actually see the value they're entering. For example. see the Margin control in the screenshot below:
- use the keyboard to focus one of the input fields
- the tooltip appears and partially hides the input
- type a value or use arrow keys to change the value: it's not possible to actually see the entered value

Screenshot 2022-07-04 at 14 08 03

On this branch, the tooltip position is OK:

Screenshot 2022-07-04 at 14 22 25

**Comment by andrewserong** [comment]
Thanks for taking a look (and for the reminder about this PR) @afercia 😀

I'll aim to dust this PR off and rebase it this week to see if we can progress it toward being mergable. Cheers!

**Comment by andrewserong** [comment]
I've given this a rebase and reverted some of the positioning logic to see if we can simplify the PR a bit. I'm not sure if it's _quite_ right yet, so will do some more testing tomorrow.

File: packages/components/src/tooltip/index.js
Comment (ntsekouras) [inline packages/components/src/tooltip/index.js:244]: Why we need this offset to be `4`?

File: packages/components/src/tooltip/index.js
Comment (ntsekouras) [reply packages/components/src/tooltip/index.js]: We need the first child's ref now because `cloneElement` would previously preserve the `ref`, correct? Also can you explain a bit why would we need the `childRef` in `getRegularElement`? Just trying to understand it fully 😄

**Review by ntsekouras — COMMENTED** [review]
Thanks for the PR @andrewserong! This tests well for me!

File: packages/components/src/tooltip/index.js
Comment (andrewserong) [reply packages/components/src/tooltip/index.js:244]: I was a bit unsure of this one, too. Prior to this PR the offset is set to `12`, I think probably to account for the positioning not being correct. With this PR, now that the Tooltip is anchored to the element, the Tooltip lines up right next to the element, so I wasn't sure if it should have some breathing room. Perhaps it's better for us to remove the `offset` in this PR, and then in a potential follow up we could expose an `offset` option for Tooltips that want to add more room?

Comment (andrewserong) [reply packages/components/src/tooltip/index.js:244]: Here's the difference between offset `4` and leaving it `undefined`:

| Offset set to 4 | No offset |
| --- | --- |
| image | image |

After testing removing the offset, I kinda prefer the extra breathing room, but happy to get rid of it if it feels too opinionated!

File: packages/components/src/tooltip/index.js
Comment (andrewserong) [reply packages/components/src/tooltip/index.js]: This one was the result of trial and error so I'm not 100% sure how it's working but here's my current understanding:

* We need a stable ref to the first child to be passed to the Popover so that it can calculate the correct position. In order to pass it reliably to the Popover without causing any errors, by using an additional `childRef` we have the one reference that is pointing to the current child element.
* Because we can't know whether or not the first child of the Tooltip is already using a `ref` for another reason, we need to attempt to merge it in with the `childRef` we're using for the `anchorRef`, this way we don't accidentally overwrite a `ref` used for other purposes.
* In `getRegularElement` and `getDisabledElement` we specify the mergedRef so that both refs are attached correctly to the cloned element (any existing ref is preserved, and our additional `childRef` for the Popover can be used for position calculation)

Let me know if you think any changes are needed here, or if we can make it clearer what's going on in the inline comment!

**Comment by andrewserong** [comment]
Thanks for taking this for a spin @ntsekouras!

I think this PR is working pretty well at the moment. My one hesitation is about the Tooltip (and Popover) position &rarr; placement calculation. I've updated the Storybook example in this PR to highlight the problem, which is that the Popover's logic which translates `left` to `-end` and `right` to `-start` works well if the Tooltip is wider than the target element, and poorly if the Tooltip is narrower than the target element.

Here's a Gif of the Storybook example with the Tooltip position set to `bottom left`:



The URL for that example (if you're running `npm run storybook:dev`) is: http://localhost:50240/?path=/story/components-tooltip--default&knob-Delay=700&knob-Position=bottom%20left&knob-Text=More%20information

I've tried hacking around a little bit with not much luck, but I _think_ ideally the Popover would be aware enough to switch the placement to the correct side based on the relative size of the Popover versus its `ref`, but I haven't been able to come up with anything that isn't super hacky just yet. It's a little tricky to come up with the right logic, too, given that the [Floating UI](https://floating-ui.com/) placement concept doesn't quite match conceptually to the existing position strings 🤔 

I'm curious if you (or others) think it's worth getting this PR in as-is, even though the positioning is reversed when the Tooltip is narrower than its `ref`, or if we should see if we can fix the position &rarr; placement logic in the Popover first?

File: packages/components/src/tooltip/index.js
Comment (ntsekouras) [reply packages/components/src/tooltip/index.js]: Thanks for explaining! I'm afraid I can't suggest something right now, since I tried doing what I asked above, that is pass only the first child's ref if exists to `cloneElement` and only pass `childRef` var as the `anchorRef` without success.. It would be great though if we were 100% sure for the solution --cc @ellatrix @diegohaz if they have any insights.

File: packages/components/src/tooltip/index.js
Comment (ntsekouras) [reply packages/components/src/tooltip/index.js:244]: I think we can keep this for now too.

**Comment by ntsekouras** [comment]
>I'm curious if you (or others) think it's worth getting this PR in as-is, even though the positioning is reversed when the Tooltip is narrower than its ref, or if we should see if we can fix the position → placement logic in the Popover first?

I can't see any difference regarding the width of the ref 🤔 . I can see the issue for your concern though:

>However, Tooltips, being quite small, have a slightly different expectation associated with them. I think for a Tooltip, when we say "bottom left" we probably mean something closer to "center the Tooltip at the bottom left corner of this element".

Is it something more? Was the position worked like this before the Popover refactor? I don't remember..

IMO this is a big improvement from trunk and I haven't observed any regression.

File: packages/components/src/tooltip/index.js
Comment (ellatrix) [inline packages/components/src/tooltip/index.js]: Can you not use `child` here? That would make things a bit clearer.

File: packages/components/src/tooltip/index.js
Comment (ellatrix) [reply packages/components/src/tooltip/index.js]: I'm having a hard time understanding what's going on here, but it's definitely good practice to create new refs for each purpose and not try to reuse refs created elsewhere. So the use of `useMergeRefs` seems fine here, if that's what you're asking.
Seeing `Children.toArray( children )[ 0 ]?.ref` makes me frown a little though. Can't you use the `child` variable that you have as well?

Comment (andrewserong) [reply packages/components/src/tooltip/index.js]: > So the use of useMergeRefs seems fine here, if that's what you're asking.

Thanks for confirming, I think that gets at the main issue?

> Seeing Children.toArray( children )[ 0 ]?.ref makes me frown a little though. Can't you use the child variable that you have as well?

This is an interesting one — I had a go at consolidating with the exiting `child` variable, but it results in a few subtle issues:
* React's `Children.only()` function [throws an error](https://reactjs.org/docs/react-api.html#reactchildrenonly) rather than returning `undefined` so we can't use it before the conditional to check how many children there are. 
* Because we're using a hook here, we have the call to the hook after a conditional.
* If we switch `child` to use the `Children.toArray( children )[ 0 ]` approach, then an incorrect `key` gets copied when the element is duplicated, which causes other components' tests to fail (I found that the hard way in https://github.com/WordPress/gutenberg/pull/41268/commits/90f5541074b829d40000979bb18d3ffcba925043)

So, as far as I can tell the best trade-off we have so far is this slightly awkward way of accessing the child in order to retrieve the existing `ref` (if it exists). I've tidied this up a tiny bit by storing the value in `const existingChildRef` to see if that helps with clarity.

Let me know if either of you can think of a better way to handle this, I've very rarely had to use `React.Children` directly so I am a little unfamiliar with this part of the API. I'm happy to keep iterating now or in follow-ups.

File: packages/components/src/tooltip/index.js
Comment (andrewserong) [reply packages/components/src/tooltip/index.js]: Replied in https://github.com/WordPress/gutenberg/pull/41268#discussion_r920743047

**Comment by andrewserong** [comment]
Thanks for the continued reviews @ntsekouras and @ellatrix!

> I can't see any difference regarding the width of the ref 🤔 

I probably did a bad job at explaining it, it's such a subtle/nuanced issue it's difficult to describe! 😄
Currently the Popover makes no change based on the width of the ref — but I was wondering if it should in order to more accurately interpret "position" in the sense of pre-Popover-refactor to "placement" after the refactor.

> Is it something more? Was the position worked like this before the Popover refactor? I don't remember..

Previously the Tooltip (and Popover) position situated the Tooltip at the desired location, rather than defining the direction in which it extends. So `bottom left` meant, "place the Tooltip at the bottom left position" rather than "extend the Tooltip to the bottom left direction". Tricky stuff! Particularly now that it's hard to find an example 😅

The issue technically exists prior to this PR, but is made more visible for the Tooltip now that it's anchored to an element rather than a zero width `span`.

It's a little subtle to work with when it comes to the Tooltip, so it might very well be something for us to dig into in follow-ups. Personally I think this PR improves things, too, so if everyone agrees, we could always land this PR to fix the immediate positioning of Tooltips issue, and continue exploring positioning/placements issues in follow-ups.

**Review by noisysocks — APPROVED** [review]
I ran into this bug while working on https://github.com/WordPress/gutenberg/pull/42352 and can confirm that it miraculously went away after merging this branch into mine.

It's a hell of a lot better than what is in `trunk`. I say let's merge it sooner than later.

PS nice work @andrewserong!

**Comment by andrewserong** [comment]
Thanks for reviewing @noisysocks! 🙇

I think merging sooner rather than later, and then seeing if we can tweak positioning (or code quality) in follow-ups sounds like a good way to go, too. @ntsekouras and @ellatrix let me know if there are any objections to merging, otherwise I'll merge this one tomorrow morning my time (just to give folks in other timezones a chance to weigh in, since there was a bit of discussion about the approach of accessing React children).

**Review by ntsekouras — APPROVED** [review]
I think it's okay to land. Thanks for your work Andrew! 💯

**Pull request merged**
